hbase-server/src/test/java/org/apache/hadoop/hbase/master/TestMasterShutdown.java

```diff
@@ -128,41 +142,50 @@ public void testMasterShutdownBeforeStartingAnyRegionServer() throws Exception {
     conf.setInt(ServerManager.WAIT_ON_REGIONSERVERS_MINTOSTART, 1);
 
     // Start the cluster
-    final HBaseTestingUtility util = new HBaseTestingUtility(conf);
-    util.startMiniDFSCluster(3);
-    util.startMiniZKCluster();
-    util.createRootDir();
-    final LocalHBaseCluster cluster =
-        new LocalHBaseCluster(conf, NUM_MASTERS, NUM_RS, HMaster.class,
-            MiniHBaseCluster.MiniHBaseClusterRegionServer.class);
-    final int MASTER_INDEX = 0;
-    final MasterThread master = cluster.getMasters().get(MASTER_INDEX);
-    master.start();
-    LOG.info("Called master start on " + master.getName());
-    Thread shutdownThread = new Thread("Shutdown-Thread") {
-      @Override
-      public void run() {
-        LOG.info("Before call to shutdown master");
-        try (Connection connection = createConnection(util); Admin admin = connection.getAdmin()) {
-          admin.shutdown();
-        } catch (Exception e) {
-          LOG.info("Error while calling Admin.shutdown, which is expected: " + e.getMessage());
+    LocalHBaseCluster cluster = null;
+    try {
+      htu = new HBaseTestingUtility(conf);
+      htu.startMiniDFSCluster(3);
+      htu.startMiniZKCluster();
+      htu.createRootDir();
+      cluster = new LocalHBaseCluster(conf, NUM_MASTERS, NUM_RS, HMaster.class,
+        MiniHBaseCluster.MiniHBaseClusterRegionServer.class);
+      final int MASTER_INDEX = 0;
+      final MasterThread master = cluster.getMasters().get(MASTER_INDEX);
+      master.start();
+      LOG.info("Called master start on " + master.getName());
+      final LocalHBaseCluster finalCluster = cluster;
+      Thread shutdownThread = new Thread("Shutdown-Thread") {
+        @Override
+        public void run() {
+          LOG.info("Before call to shutdown master");
+          try (Connection connection = createConnection(htu); Admin admin = connection.getAdmin()) {
+            admin.shutdown();
```

Comment:
This is the place that the master registry exposed a race ..(shutdown goes missing..). Rebase will not be clean now :'(

Reply:
I think this is the source of the instability I've seen in this test, but I cannot reproduce it reliably.

You see this test failing reliably on your branch?

Reply:
No, not after adding this check https://github.com/apache/hbase/blob/d110c08dce482ef3161294ab931d2559d8e57fca/hbase-server/src/test/java/org/apache/hadoop/hbase/master/TestMasterShutdown.java#L141

This check basically works around the actual problem without fixing it. Ideally you could do the same because that check is anyway applied in the master after the branch merge.

Fwiw, I think this race is exposed after committing HBASE-23764, because that speeds up the connections.